src/main/java/teammates/ui/webapi/action/GetStudentProfileAction.java

```diff
@@ -16,15 +18,55 @@ protected AuthType getMinAuthLevel() {
 
     @Override
     public void checkSpecificAccessControl() {
-        if (!userInfo.isStudent) {
-            throw new UnauthorizedAccessException("Student privilege is required to access this resource.");
+
+        String entityType = getNonNullRequestParamValue(Const.ParamsNames.ENTITY_TYPE);
+        if (userInfo.isStudent && entityType.equals(Const.EntityType.STUDENT)) {
+            checkStudentAccess();
+        } else if (userInfo.isInstructor && entityType.equals(Const.EntityType.INSTRUCTOR)) {
+            checkInstructorAccess();
+        } else {
+            throw new UnauthorizedAccessException("Your account cannot access this student's profile.");
+        }
+    }
+
+    private void checkStudentAccess() {
+        String studentId = getNonNullRequestParamValue(Const.ParamsNames.STUDENT_ID);
+        if (studentId.equals(userInfo.id)) {
+            // Student wants to access his own profile, permission always granted.
+            return;
         }
+        String courseId = getNonNullRequestParamValue(Const.ParamsNames.COURSE_ID);
+
+        StudentAttributes studentUser = logic.getStudentForGoogleId(courseId, userInfo.id);
+        if (studentUser == null) {
+            throw new UnauthorizedAccessException("You are not the classmate of this student,"
```

Comment:
This message (and the others) - are they intended to be shown to the user in the case of API failure? Perhaps something like "You are this student's classmate, so you cannot access their profile" would be more appropriate?

Reply:
That's contrary to the fact. The logic here is that if the user account is not found in the courseId given in the parameter, he is not the classmate of the student given by his Id